kubernetes/provider.go

```diff
@@ -225,6 +228,13 @@ func (k kubeClientsets) MainClientset() (*kubernetes.Clientset, error) {
 	return k.mainClientset, nil
 }
 
+func (k kubeClientsets) GetContext() (*context.Context, error) {
```

Comment:
I'm questioning the value of having this method, and storing a shared context here alongside the clients and passing down a pointer to it. I know that at this point it is academic as the context doesn't actually do anything yet.

It more makes sense that each CRUD operation would happen in its own context, or receive its context from the calling function in the case where a Create immediately then does a Read. This is actually what's going to happen when we migrate to SDK v2 - each CRUD func is going to have a context aware version that will supply a `context.Context` as the first argument.

See here: https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#more-support-for-context-context

What I would do here is add a `ctx := context.TODO()` to the top of every CRUD func, then we would simply remove that line when we do the SDK migration as `ctx`will become the first argument of the function.

cc @alexsomesan for his $0.02 on this.

Sorry, I know you've added this in like 200 places 😱

Reply:
Yeah, I was feeling uncertain about that function too. I know we don't want to share a context too widely. Here is some information on it, for reference:

https://github.com/golang/go/issues/22602
```
Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx: [...]"

"While we've told people not to add contexts to structs, I think that guidance is over-aggressive. The real advice is not to store contexts. They should be passed along like parameters. But if the struct is essentially just a parameter, it's okay. I think this concern can be addressed with package-level documentation and examples."
```

I thought that by storing it in that struct, it would allow us to share a context when needed, and we would have the option to declare new contexts where those are required too. So +1 to not using it everywhere. I'm just not sure where to draw that line. Some functions should probably share a context, while others should not. Maybe adding one per CRUD operation is the way to go, like you said.

Reply:
I think John has a good point here. Taking a step back to think about the role of the context in our case, it's to coordinate the cancelation of upstream connections in case the "parent" operation (apply, destroy) is stopped prematurely for whatever reason. So we should always be passing around the context corresponding to that specific operation. Passing any other (older) context isn't adding value for the effort required to use them.

That being said, I second what John suggested about just using `ctx := context.TODO()` in every CRUD until we gain access to the new SDKs context aware interface.